Add a public export map accessor for unit testing (NodeJS)

i-am-tom opened the pull request:
Continuation of https://github.com/pulumi/pulumi/pull/20472 to address https://github.com/pulumi/pulumi/issues/6113.

Long-time listeners may remember this PR from before, and our conversations to avoid it using the event stream. While I still think that's a better fix, this issue needs closing, and the code already exists in Go, so we should probably just add this for now and deprecate it when we have a better solution.

Frassle commented on sdk/nodejs/runtime/resource.ts:
```diff
@@ -1177,6 +1177,9 @@ export function registerResourceOutputs(res: Resource, outputs: Inputs | Promise
                 );
 
+                const store = getStore();
+                store.currentExportMap = resolved.outputs;
```
We're writing this for every output registration? Doesn't that mean we could end up with any component resource being the last due to running inside an apply and so not actually getting the stack outputs here. I think this needs to be moved to the stack class or protected here at least by checking the type name.

i-am-tom committed:
Add a public export map accessor for unit testing (NodeJS)

Continuation of #20471 for #6113.

Adds `currentExportMap` to the runtime store and populates it from
`Stack.initialize` after stack outputs are resolved. Lets unit tests
inspect what a program exported without running it under a real engine.

Tests cover that the map captures resolved outputs and is replaced
(not merged) on each sequential `runInPulumiStack` call.

i-am-tom commented on sdk/nodejs/tests_with_mocks/exports.spec.ts:
```diff
@@ -0,0 +1,44 @@
+
+        assert.deepStrictEqual(outs, { a: 42, b: 99, c: { x: { n: 1 }, y: "z" } });
+
+        const exportMap = getStore().currentExportMap as Record<string, any> | undefined;
```
OK, I did some digging and the answer is "no, but":

- This is already a problem with `setStackResource`, and we already have tests that use this.
- Fixing both this and the race conditions would require a pretty drastic overhaul of the tests (we can't just have a local storage per-test because we use e.g. `setMocks` in a `before` in various places, so we can't encapsulate that _and_ the test in the same closure.

This may well be why we don't already run the node tests in parallel (they're currently all sequential) - in any case, this test shouldn't flake because we won't parallelise, so we won't race.

Frassle reviewed (commented):
I do not like this. I think people will write horrible code around this. We already return the stack outputs from pulumi.runtime.runInPulumiStack, can't tests just use that?
